python/hawkey/query-py.c

```diff
@@ -303,7 +396,127 @@ filter(_QueryObject *self, PyObject *args)
         break;
     }
     }
-    Py_RETURN_NONE;
+    return 1;
+}
+
+static char *
+filter_key_splitter(char** key)
+{
+    char *sbegin = *key;
+	char *end;
+
+    if (sbegin == NULL)
+		return NULL;
+    int index;
+
+    for (index = 0; sbegin[index] != '\0'; ++index) {
+        if ((sbegin[index] == '_') &&  (sbegin[index + 1] == '_')) {
+            end = sbegin + index;
+            *end++ = '\0';
+            *key = ++end;
+            return sbegin;
+        }
+    }
+    *key = NULL;
+    return sbegin;
+}
+
+static PyObject *
+filter_internal(HyQuery query, PyObject *sack, PyObject *args, PyObject *kwds)
+{
+    PyObject *key, *value;
+    Py_ssize_t pos = 0;
+    key_t keyname;
+    int cmp_type;
+    PyObject *tuple_item;
+    const char *cmatch; 
+    char *parsed_string, *parcial_string;
+    int argument_number, presence_cmp_type;
+    int cmp_type_flag = 0;
+
+    if (args != NULL) {
+        Py_ssize_t tuple_size = PyTuple_Size(args);
+        for (int x = 0; x < tuple_size; ++x) {
+            tuple_item = PyTuple_GetItem(args, x);
+            if (PyInt_Check(tuple_item)) {
+                long c_int = PyLong_AsLong(tuple_item);
+                if (c_int == HY_ICASE) {
+                    cmp_type_flag = HY_ICASE;
+                } else {
+                    PyErr_SetString(PyExc_ValueError, "Invalid flag. Only HY_ICASE allowed");
+                    Py_RETURN_NONE;
+                }
+            }
+        }
+    }
+
+    if (kwds != NULL) {
+        while (PyDict_Next(kwds, &pos, &key, &value)) {
+            keyname = -1;
+            argument_number = 0;
+            PyObject *tmp_py_str = NULL;
+            cmatch = pycomp_get_string(key, &tmp_py_str);
+            parsed_string = strdup(cmatch);
```

Comment:
1. Memory leak. strdup() allocates new memory. Yes you need it.  But where do you release the memory?
2. You can call `Py_XDECREF(tmp_py_str);` after this line instead of calling it before each `return`.

Reply:
Probably simplest solution:
Remove declaration of parsed_string from line 433 and do `g_autofree char *parsed_string = strdup(cmatch);`

Reply:
filter_key_splitter() modifies its argument. So using temporary string is solution:
```
g_autofree char *parsed_string = strdup(cmatch);
char *tmp_string = parsed_string;
cmp_type = 0;
while ((parcial_string = filter_key_splitter(&tmp_string)) != NULL) {
```